File: post_montecarlo/montecarlo_python_marketing.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import warnings
warnings.filterwarnings('ignore')

# Configuração para melhor visualização
plt.style.use('seaborn-v0_8-darkgrid')
sns.set_palette("husl")

def criar_parametros_canais():
    """
    Define os parâmetros de cada canal de marketing com suas incertezas
    
    Retorna:
    DataFrame com parâmetros de cada canal
    """
    canais = {
        'canal': ['Google Ads', 'Facebook Ads', 'Instagram Ads', 'Email Marketing', 'LinkedIn Ads'],
        'cpc_medio': [2.5, 1.8, 1.5, 0.05, 5.0],  # Custo por clique médio
        'cpc_desvio': [0.5, 0.4, 0.3, 0.01, 1.0],  # Desvio padrão do CPC
        'ctr_medio': [0.035, 0.025, 0.022, 0.15, 0.028],  # Taxa de clique média
        'ctr_desvio': [0.008, 0.006, 0.005, 0.03, 0.007],  # Desvio da CTR
        'conversao_media': [0.03, 0.025, 0.02, 0.05, 0.04],  # Taxa de conversão média
        'conversao_desvio': [0.008, 0.007, 0.005, 0.015, 0.01],  # Desvio da conversão
        'ticket_medio': [150, 120, 100, 200, 300],  # Valor médio da venda
        'ticket_desvio': [30, 25, 20, 50, 80]  # Desvio do ticket médio
    }
    
    return pd.DataFrame(canais)

def simular_canal(nome_canal, budget, parametros, n_simulacoes=10000):
    """
    Simula o desempenho de um canal de marketing usando Monte Carlo
    
    Parâmetros:
    nome_canal: nome do canal a simular
    budget: orçamento alocado para o canal
    parametros: DataFrame com parâmetros dos canais
    n_simulacoes: número de simulações
    
    Retorna:
    DataFrame com resultados das simulações
    """
    # Filtra parâmetros do canal
    canal = parametros[parametros['canal'] == nome_canal].iloc[0]
    
    # Arrays para armazenar resultados
    resultados = {
        'simulacao': range(n_simulacoes),
        'cpc': np.zeros(n_simulacoes),
        'ctr': np.zeros(n_simulacoes),
        'taxa_conversao': np.zeros(n_simulacoes),
        'ticket_medio': np.zeros(n_simulacoes),
       'cliques': np.zeros(n_simulacoes),
       'conversoes': np.zeros(n_simulacoes),
       'receita': np.zeros(n_simulacoes),
       'roi': np.zeros(n_simulacoes),
       'lucro': np.zeros(n_simulacoes)
   }
   
   for i in range(n_simulacoes):
       # Simula variáveis com distribuição normal truncada (valores positivos)
       cpc = max(0.01, np.random.normal(canal['cpc_medio'], canal['cpc_desvio']))
       ctr = np.clip(np.random.normal(canal['ctr_medio'], canal['ctr_desvio']), 0.001, 1)
       conversao = np.clip(np.random.normal(canal['conversao_media'], canal['conversao_desvio']), 0.001, 1)
       ticket = max(10, np.random.normal(canal['ticket_medio'], canal['ticket_desvio']))
       
       # Calcula métricas
       cliques = int(budget / cpc)
       impressoes = int(cliques / ctr)
       conversoes = int(cliques * conversao)
       receita = conversoes * ticket
       lucro = receita - budget
       roi = (lucro / budget) * 100 if budget > 0 else 0
       
       # Armazena resultados
       resultados['cpc'][i] = cpc
       resultados['ctr'][i] = ctr
       resultados['taxa_conversao'][i] = conversao
       resultados['ticket_medio'][i] = ticket
       resultados['cliques'][i] = cliques
       resultados['conversoes'][i] = conversoes
       resultados['receita'][i] = receita
       resultados['roi'][i] = roi
       resultados['lucro'][i] = lucro
   
   df_resultados = pd.DataFrame(resultados)
   df_resultados['canal'] = nome_canal
   
   return df_resultados

def analisar_resultados_canal(df_simulacao, canal_nome):
   """
   Cria visualizações e estatísticas dos resultados de simulação
   """
   fig, axes = plt.subplots(2, 2, figsize=(15, 10))
   fig.suptitle(f'Análise de Monte Carlo - {canal_nome}', fontsize=16)
   
   # Distribuição do ROI
   ax1 = axes[0, 0]
   ax1.hist(df_simulacao['roi'], bins=50, alpha=0.7, color='blue', edgecolor='black')
   ax1.axvline(df_simulacao['roi'].mean(), color='red', linestyle='--', 
               label=f'Média: {df_simulacao["roi"].mean():.1f}%')
   ax1.set_xlabel('ROI (%)')
   ax1.set_ylabel('Frequência')
   ax1.set_title('Distribuição do ROI')
   ax1.legend()
   
   # Distribuição do Lucro
   ax2 = axes[0, 1]
   ax2.hist(df_simulacao['lucro'], bins=50, alpha=0.7, color='green', edgecolor='black')
   ax2.axvline(df_simulacao['lucro'].mean(), color='red', linestyle='--',
               label=f'Média: R$ {df_simulacao["lucro"].mean():.2f}')
   ax2.set_xlabel('Lucro (R$)')
   ax2.set_ylabel('Frequência')
   ax2.set_title('Distribuição do Lucro')
   ax2.legend()
   
   # Análise de Risco (Probabilidade de Prejuízo)
   ax3 = axes[1, 0]
   prob_prejuizo = (df_simulacao['lucro'] < 0).mean() * 100
   prob_lucro = (df_simulacao['lucro'] >= 0).mean() * 100
   
   ax3.bar(['Prejuízo', 'Lucro'], [prob_prejuizo, prob_lucro], 
           color=['red', 'green'], alpha=0.7)
   ax3.set_ylabel('Probabilidade (%)')
   ax3.set_title('Análise de Risco')
   for i, v in enumerate([prob_prejuizo, prob_lucro]):
       ax3.text(i, v + 1, f'{v:.1f}%', ha='center', fontweight='bold')
   
   # Box plot de métricas chave
   ax4 = axes[1, 1]
   metricas = df_simulacao[['conversoes', 'receita', 'lucro']].copy()
   metricas['receita'] = metricas['receita'] / 100  # Escala para melhor visualização
   metricas['lucro'] = metricas['lucro'] / 100
   
   metricas.boxplot(ax=ax4)
   ax4.set_title('Variabilidade das Métricas (Receita e Lucro /100)')
   ax4.set_ylabel('Valor')
   
   plt.tight_layout()
   plt.show()
   
   # Estatísticas resumidas
   print(f"\n=== Estatísticas para {canal_nome} ===")
   print(f"ROI Médio: {df_simulacao['roi'].mean():.2f}%")
   print(f"ROI Mediano: {df_simulacao['roi'].median():.2f}%")
   print(f"Desvio Padrão ROI: {df_simulacao['roi'].std():.2f}%")
   print(f"Probabilidade de Prejuízo: {prob_prejuizo:.1f}%")
   print(f"Value at Risk (5%): R$ {df_simulacao['lucro'].quantile(0.05):.2f}")
   print(f"Lucro Esperado: R$ {df_simulacao['lucro'].mean():.2f}")

def otimizar_portfolio(budget_total, parametros, n_simulacoes=5000):
   """
   Encontra a melhor alocação de budget entre canais usando Monte Carlo
   """
   canais = parametros['canal'].values
   n_canais = len(canais)
   
   # Gera alocações aleatórias
   n_portfolios = 1000
   alocacoes = np.random.dirichlet(np.ones(n_canais), size=n_portfolios)
   
   resultados_portfolio = []
   
   for i, alocacao in enumerate(alocacoes):
       roi_total = 0
       lucro_total = 0
       risco_total = 0
       
       # Simula cada canal com sua alocação
       for j, canal in enumerate(canais):
           budget_canal = budget_total * alocacao[j]
           if budget_canal > 0:
               sim = simular_canal(canal, budget_canal, parametros, n_simulacoes=1000)
               roi_medio = sim['roi'].mean()
               lucro_medio = sim['lucro'].mean()
               risco = sim['lucro'].std()
               
               # Ponderação pelo budget
               roi_total += roi_medio * alocacao[j]
               lucro_total += lucro_medio
               risco_total += risco * alocacao[j]
       
       resultados_portfolio.append({
           'portfolio': i,
           'roi_esperado': roi_total,
           'lucro_esperado': lucro_total,
           'risco': risco_total,
           'sharpe_ratio': lucro_total / risco_total if risco_total > 0 else 0,
           'alocacao': alocacao
       })
   
   df_portfolios = pd.DataFrame(resultados_portfolio)
   
   # Encontra portfolios ótimos
   melhor_roi = df_portfolios.loc[df_portfolios['roi_esperado'].idxmax()]
   melhor_sharpe = df_portfolios.loc[df_portfolios['sharpe_ratio'].idxmax()]
   
   return df_portfolios, melhor_roi, melhor_sharpe

def plotar_fronteira_eficiente(df_portfolios, melhor_roi, melhor_sharpe):
   """
   Plota a fronteira eficiente de portfolios de marketing
   """
   plt.figure(figsize=(12, 8))
   
   # Scatter plot de todos os portfolios
   scatter = plt.scatter(df_portfolios['risco'], 
                        df_portfolios['lucro_esperado'],
                        c=df_portfolios['sharpe_ratio'], 
                        cmap='viridis', 
                        alpha=0.6,
                        s=50)
   
   # Destaca portfolios ótimos
   plt.scatter(melhor_roi['risco'], melhor_roi['lucro_esperado'], 
              color='red', s=200, marker='*', 
              label='Melhor ROI', edgecolor='black', linewidth=2)
   
   plt.scatter(melhor_sharpe['risco'], melhor_sharpe['lucro_esperado'], 
              color='orange', s=200, marker='*', 
              label='Melhor Sharpe Ratio', edgecolor='black', linewidth=2)
   
   plt.colorbar(scatter, label='Sharpe Ratio')
   plt.xlabel('Risco (Desvio Padrão do Lucro)')
   plt.ylabel('Lucro Esperado (R$)')
   plt.title('Fronteira Eficiente - Portfolios de Marketing')
   plt.legend()
   plt.grid(True, alpha=0.3)
   
   plt.tight_layout()
   plt.show()

def exibir_alocacao_otima(melhor_portfolio, parametros, titulo):
   """
   Exibe a alocação ótima de budget em um gráfico de pizza
   """
   canais = parametros['canal'].values
   alocacao = melhor_portfolio['alocacao']
   
   # Filtra canais com alocação > 1%
   mask = alocacao > 0.01
   canais_filtrados = canais[mask]
   alocacao_filtrada = alocacao[mask]
   
   plt.figure(figsize=(10, 8))
   colors = plt.cm.Set3(range(len(canais_filtrados)))
   
   wedges, texts, autotexts = plt.pie(alocacao_filtrada, 
                                      labels=canais_filtrados,
                                      colors=colors,
                                      autopct='%1.1f%%',
                                      startangle=90)
   
   plt.title(f'{titulo}\nROI Esperado: {melhor_portfolio["roi_esperado"]:.1f}% | '
             f'Lucro Esperado: R$ {melhor_portfolio["lucro_esperado"]:.2f}')
   
   plt.axis('equal')
   plt.tight_layout()
   plt.show()

# Exemplo de uso completo
if __name__ == "__main__":
   # Define parâmetros
   parametros = criar_parametros_canais()
   budget_total = 50000  # R$ 50.000
   
   # 1. Análise individual de cada canal
   print("=== ANÁLISE INDIVIDUAL DOS CANAIS ===")
   for canal in parametros['canal']:
       print(f"\nSimulando {canal}...")
       simulacao = simular_canal(canal, budget_total/5, parametros, n_simulacoes=10000)
       analisar_resultados_canal(simulacao, canal)
   
   # 2. Otimização de portfolio
   print("\n=== OTIMIZAÇÃO DE PORTFOLIO ===")
   df_portfolios, melhor_roi, melhor_sharpe = otimizar_portfolio(budget_total, parametros)
   
   # 3. Visualização da fronteira eficiente
   plotar_fronteira_eficiente(df_portfolios, melhor_roi, melhor_sharpe)
   
   # 4. Exibição das alocações ótimas
   exibir_alocacao_otima(melhor_roi, parametros, "Portfolio de Melhor ROI")
   exibir_alocacao_otima(melhor_sharpe, parametros, "Portfolio de Melhor Sharpe Ratio")
   
   # 5. Comparação de estratégias
   print("\n=== COMPARAÇÃO DE ESTRATÉGIAS ===")
   print(f"Portfolio de Melhor ROI:")
   print(f"  - ROI Esperado: {melhor_roi['roi_esperado']:.2f}%")
   print(f"  - Lucro Esperado: R$ {melhor_roi['lucro_esperado']:.2f}")
   print(f"  - Risco: R$ {melhor_roi['risco']:.2f}")
   
   print(f"\nPortfolio de Melhor Sharpe Ratio:")
   print(f"  - ROI Esperado: {melhor_sharpe['roi_esperado']:.2f}%")
   print(f"  - Lucro Esperado: R$ {melhor_sharpe['lucro_esperado']:.2f}")
   print(f"  - Risco: R$ {melhor_sharpe['risco']:.2f}")

File: post_montecarlo/montecarlo_python_marketing.R
# Carrega bibliotecas
library(tidyverse)
library(scales)
library(gridExtra)
library(viridis)

# Define parâmetros dos canais
criar_parametros_canais <- function() {
  tibble(
    canal = c('Google Ads', 'Facebook Ads', 'Instagram Ads', 'Email Marketing', 'LinkedIn Ads'),
    cpc_medio = c(2.5, 1.8, 1.5, 0.05, 5.0),
    cpc_desvio = c(0.5, 0.4, 0.3, 0.01, 1.0),
    ctr_medio = c(0.035, 0.025, 0.022, 0.15, 0.028),
    ctr_desvio = c(0.008, 0.006, 0.005, 0.03, 0.007),
    conversao_media = c(0.03, 0.025, 0.02, 0.05, 0.04),
    conversao_desvio = c(0.008, 0.007, 0.005, 0.015, 0.01),
    ticket_medio = c(150, 120, 100, 200, 300),
    ticket_desvio = c(30, 25, 20, 50, 80)
  )
}

simular_canal <- function(nome_canal, budget, parametros, n_simulacoes = 10000) {
  # Filtra parâmetros do canal
  canal <- parametros |>
    filter(canal == nome_canal) |>
    slice(1)
  
  # Vetoriza as simulações para melhor performance
  simulacoes <- tibble(
    simulacao = 1:n_simulacoes,
    # Simula variáveis com distribuição normal truncada
    cpc = pmax(0.01, rnorm(n_simulacoes, canal$cpc_medio, canal$cpc_desvio)),
    ctr = pmin(pmax(rnorm(n_simulacoes, canal$ctr_medio, canal$ctr_desvio), 0.001), 1),
    taxa_conversao = pmin(pmax(rnorm(n_simulacoes, canal$conversao_media, canal$conversao_desvio), 0.001), 1),
    ticket_medio = pmax(10, rnorm(n_simulacoes, canal$ticket_medio, canal$ticket_desvio))
  ) |>
    mutate(
      cliques = as.integer(budget / cpc),
      impressoes = as.integer(cliques / ctr),
      conversoes = as.integer(cliques * taxa_conversao),
      receita = conversoes * ticket_medio,
      lucro = receita - budget,
      roi = if_else(budget > 0, (lucro / budget) * 100, 0),
      canal = nome_canal
    )
  
  return(simulacoes)
}

analisar_resultados_canal <- function(df_simulacao, canal_nome) {
  # Calcula estatísticas
  stats <- df_simulacao |>
    summarise(
      roi_medio = mean(roi),
      roi_mediano = median(roi),
      roi_desvio = sd(roi),
      prob_prejuizo = mean(lucro < 0) * 100,
      var_5pct = quantile(lucro, 0.05),
      lucro_esperado = mean(lucro)
    )
  
  # Gráfico 1: Distribuição do ROI
  p1 <- ggplot(df_simulacao, aes(x = roi)) +
    geom_histogram(bins = 50, fill = "blue", alpha = 0.7, color = "black") +
    geom_vline(xintercept = stats$roi_medio, color = "red", 
               linetype = "dashed", size = 1) +
    labs(
      title = "Distribuição do ROI",
      x = "ROI (%)",
      y = "Frequência"
    ) +
    annotate("text", x = stats$roi_medio + 10, y = Inf, 
             label = sprintf("Média: %.1f%%", stats$roi_medio),
             vjust = 2, hjust = 0, color = "red")
  
  # Gráfico 2: Distribuição do Lucro
  p2 <- ggplot(df_simulacao, aes(x = lucro)) +
    geom_histogram(bins = 50, fill = "green", alpha = 0.7, color = "black") +
    geom_vline(xintercept = stats$lucro_esperado, color = "red", 
               linetype = "dashed", size = 1) +
    labs(
      title = "Distribuição do Lucro",
      x = "Lucro (R$)",
      y = "Frequência"
    ) +
    scale_x_continuous(labels = scales::dollar_format(prefix = "R$ "))
  
  # Gráfico 3: Análise de Risco
  risco_data <- tibble(
    categoria = c("Prejuízo", "Lucro"),
    probabilidade = c(stats$prob_prejuizo, 100 - stats$prob_prejuizo)
  )
  
  p3 <- ggplot(risco_data, aes(x = categoria, y = probabilidade, fill = categoria)) +
    geom_bar(stat = "identity", alpha = 0.7) +
    geom_text(aes(label = sprintf("%.1f%%", probabilidade)), 
              vjust = -0.5, fontface = "bold") +
    scale_fill_manual(values = c("Prejuízo" = "red", "Lucro" = "green")) +
    labs(
      title = "Análise de Risco",
      y = "Probabilidade (%)"
    ) +
    theme(legend.position = "none")
  
  # Gráfico 4: Boxplot de métricas
  metricas_long <- df_simulacao |>
    select(conversoes, receita, lucro) |>
    mutate(
      receita = receita / 100,
      lucro = lucro / 100
    ) |>
    pivot_longer(everything(), names_to = "metrica", values_to = "valor")
  
  p4 <- ggplot(metricas_long, aes(x = metrica, y = valor)) +
    geom_boxplot(fill = "lightblue", alpha = 0.7) +
    labs(
      title = "Variabilidade das Métricas",
      y = "Valor (Receita e Lucro /100)"
    )
  
  # Combina os gráficos
  grid.arrange(p1, p2, p3, p4, ncol = 2, 
               top = paste("Análise de Monte Carlo -", canal_nome))
  
  # Imprime estatísticas
  cat("\n=== Estatísticas para", canal_nome, "===\n")
  cat("ROI Médio:", sprintf("%.2f%%", stats$roi_medio), "\n")
  cat("ROI Mediano:", sprintf("%.2f%%", stats$roi_mediano), "\n")
  cat("Desvio Padrão ROI:", sprintf("%.2f%%", stats$roi_desvio), "\n")
  cat("Probabilidade de Prejuízo:", sprintf("%.1f%%", stats$prob_prejuizo), "\n")
  cat("Value at Risk (5%):", sprintf("R$ %.2f", stats$var_5pct), "\n")
  cat("Lucro Esperado:", sprintf("R$ %.2f", stats$lucro_esperado), "\n")
}

otimizar_portfolio <- function(budget_total, parametros, n_simulacoes = 5000) {
  canais <- parametros$canal
  n_canais <- length(canais)
  
  # Gera alocações usando Dirichlet
  n_portfolios <- 1000
  
  # Função para gerar uma alocação Dirichlet
  gerar_alocacao <- function() {
    x <- rgamma(n_canais, 1)
    x / sum(x)
  }
  
  # Gera múltiplas alocações
  alocacoes <- replicate(n_portfolios, gerar_alocacao()) |> t()
  
  # Calcula métricas para cada portfolio
  resultados_portfolio <- 1:n_portfolios |>
    map_dfr(function(i) {
      alocacao <- alocacoes[i, ]
      
      # Simula cada canal
      metricas <- 1:n_canais |>
        map_dfr(function(j) {
          budget_canal <- budget_total * alocacao[j]
          if (budget_canal > 0) {
            sim <- simular_canal(canais[j], budget_canal, parametros, 1000)
            tibble(
              canal = canais[j],
              peso = alocacao[j],
              roi_medio = mean(sim$roi),
              lucro_medio = mean(sim$lucro),
              risco = sd(sim$lucro)
            )
          } else {
            tibble(
              canal = canais[j],
              peso = 0,
              roi_medio = 0,
              lucro_medio = 0,
              risco = 0
            )
          }
        })
      
      # Calcula métricas do portfolio
      tibble(
        portfolio = i,
        roi_esperado = sum(metricas$roi_medio * metricas$peso),
        lucro_esperado = sum(metricas$lucro_medio),
        risco = sqrt(sum((metricas$risco * metricas$peso)^2)),
        sharpe_ratio = lucro_esperado / risco,
        alocacao = list(alocacao)
      )
    })
  
  # Encontra portfolios ótimos
  melhor_roi <- resultados_portfolio |>
    slice_max(roi_esperado, n = 1)
  
  melhor_sharpe <- resultados_portfolio |>
    slice_max(sharpe_ratio, n = 1)
  
  return(list(
    portfolios = resultados_portfolio,
    melhor_roi = melhor_roi,
    melhor_sharpe = melhor_sharpe
  ))
}

plotar_fronteira_eficiente <- function(resultado_otimizacao) {
  portfolios <- resultado_otimizacao$portfolios
  melhor_roi <- resultado_otimizacao$melhor_roi
  melhor_sharpe <- resultado_otimizacao$melhor_sharpe
  
  p <- ggplot(portfolios, aes(x = risco, y = lucro_esperado)) +
    geom_point(aes(color = sharpe_ratio), size = 3, alpha = 0.6) +
    scale_color_viridis(name = "Sharpe Ratio") +
    # Destaca portfolios ótimos
    geom_point(data = melhor_roi, 
               color = "red", size = 8, shape = 17) +
    geom_point(data = melhor_sharpe, 
               color = "orange", size = 8, shape = 17) +
    # Adiciona labels
    annotate("text", x = melhor_roi$risco, y = melhor_roi$lucro_esperado,
             label = "Melhor ROI", vjust = -1, color = "red", fontface = "bold") +
    annotate("text", x = melhor_sharpe$risco, y = melhor_sharpe$lucro_esperado,
             label = "Melhor Sharpe", vjust = -1, color = "orange", fontface = "bold") +
    labs(
      title = "Fronteira Eficiente - Portfolios de Marketing",
      x = "Risco (Desvio Padrão do Lucro)",
      y = "Lucro Esperado (R$)"
    ) +
    theme_minimal() +
    theme(
      plot.title = element_text(hjust = 0.5, size = 16, face = "bold"),
      legend.position = "right"
    )
  
  print(p)
}

exibir_alocacao_otima <- function(melhor_portfolio, parametros, titulo) {
  canais <- parametros$canal
  alocacao <- melhor_portfolio$alocacao[[1]]
  
  # Cria data frame para o gráfico
  dados_pizza <- tibble(
    canal = canais,
    alocacao = alocacao * 100
  ) |>
    filter(alocacao > 1) |>  # Filtra canais com alocação > 1%
    arrange(desc(alocacao))
  
  # Gráfico de pizza
  p <- ggplot(dados_pizza, aes(x = "", y = alocacao, fill = canal)) +
    geom_bar(stat = "identity", width = 1) +
    coord_polar("y", start = 0) +
    geom_text(aes(label = sprintf("%.1f%%", alocacao)),
              position = position_stack(vjust = 0.5)) +
    labs(
      title = paste0(titulo, "\n",
                    sprintf("ROI Esperado: %.1f%% | Lucro Esperado: R$ %.2f",
                           melhor_portfolio$roi_esperado,
                           melhor_portfolio$lucro_esperado)),
      fill = "Canal"
    ) +
    theme_void() +
    theme(
      plot.title = element_text(hjust = 0.5, size = 14, face = "bold"),
      legend.position = "right"
    )
  
  print(p)
}

# Exemplo de uso completo
# Define parâmetros
parametros <- criar_parametros_canais()
budget_total <- 50000  # R$ 50.000

# 1. Análise individual de cada canal
cat("=== ANÁLISE INDIVIDUAL DOS CANAIS ===\n")
for (canal_nome in parametros$canal) {
  cat("\nSimulando", canal_nome, "...\n")
  simulacao <- simular_canal(canal_nome, budget_total/5, parametros, n_simulacoes = 10000)
  analisar_resultados_canal(simulacao, canal_nome)
}

# 2. Otimização de portfolio
cat("\n=== OTIMIZAÇÃO DE PORTFOLIO ===\n")
resultado_otimizacao <- otimizar_portfolio(budget_total, parametros)

# 3. Visualização da fronteira eficiente
plotar_fronteira_eficiente(resultado_otimizacao)

# 4. Exibição das alocações ótimas
exibir_alocacao_otima(resultado_otimizacao$melhor_roi, parametros, "Portfolio de Melhor ROI")
exibir_alocacao_otima(resultado_otimizacao$melhor_sharpe, parametros, "Portfolio de Melhor Sharpe Ratio")

# 5. Comparação de estratégias
cat("\n=== COMPARAÇÃO DE ESTRATÉGIAS ===\n")
cat("Portfolio de Melhor ROI:\n")
cat(sprintf("  - ROI Esperado: %.2f%%\n", resultado_otimizacao$melhor_roi$roi_esperado))
cat(sprintf("  - Lucro Esperado: R$ %.2f\n", resultado_otimizacao$melhor_roi$lucro_esperado))
cat(sprintf("  - Risco: R$ %.2f\n", resultado_otimizacao$melhor_roi$risco))

cat("\nPortfolio de Melhor Sharpe Ratio:\n")
cat(sprintf("  - ROI Esperado: %.2f%%\n", resultado_otimizacao$melhor_sharpe$roi_esperado))
cat(sprintf("  - Lucro Esperado: R$ %.2f\n", resultado_otimizacao$melhor_sharpe$lucro_esperado))
cat(sprintf("  - Risco: R$ %.2f\n", resultado_otimizacao$melhor_sharpe$risco))